x/twap/api_test.go

```diff
@@ -238,9 +237,158 @@ func (s *TestSuite) TestGetArithmeticTwap() {
+
+		// baseTimePlusKeepPeriod = baseTime + defaultRecordHistoryKeepPeriod
+		baseTimePlusKeepPeriod = baseTime.Add(defaultRecordHistoryKeepPeriod)
+
+		// oneHourBeforeKeepThreshold =  baseKeepThreshold - 1 hour
+		oneHourBeforeKeepThreshold = baseTimePlusKeepPeriod.Add(-time.Hour)
+
+		// oneHourAfterKeepThreshold = baseKeepThreshold + 1 hour
+		oneHourAfterKeepThreshold = baseTimePlusKeepPeriod.Add(time.Hour)
+
+		periodBetweenBaseAndOneHourBeforeThreshold           = (defaultRecordHistoryKeepPeriod.Milliseconds() - time.Hour.Milliseconds())
+		accumBeforeKeepThreshold0, accumBeforeKeepThreshold1 = sdk.NewDec(periodBetweenBaseAndOneHourBeforeThreshold * 10), sdk.NewDec(periodBetweenBaseAndOneHourBeforeThreshold * 10)
+		// recordBeforeKeepThreshold is a record with t=baseTime+keepPeriod-1h, sp0=30(sp1=0.3) accumulators set relative to baseRecord
+		recordBeforeKeepThreshold types.TwapRecord = newTwapRecordWithDefaults(oneHourBeforeKeepThreshold, sdk.NewDec(30), accumBeforeKeepThreshold0, accumBeforeKeepThreshold1)
+	)
+
+	// N.B.: when ctxTime = end time, we trigger the "TWAP to now path".
+	// As a result, we duplicate the test cases by trigerring both "to now" and "with end time" paths
+	// To trigger "with end time" path, we make end time less than ctxTime.
+	tests := map[string]struct {
+		recordsToSet []types.TwapRecord
+		ctxTime      time.Time
+		input        getTwapInput
+		expTwap      sdk.Dec
+		expErrorStr  string
+	}{
+		"(1 record at keep threshold); to now; ctxTime = at keep threshold; start time = end time = base keep threshold": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      baseTimePlusKeepPeriod,
+			input:        makeSimpleTwapInput(baseTimePlusKeepPeriod, baseTimePlusKeepPeriod, quoteAssetA),
+			expTwap:      sdk.NewDec(10),
+		},
+		"(1 record at keep threshold); with end time; ctxTime = at keep threshold; start time = end time = base keep threshold - 1ms": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      baseTimePlusKeepPeriod,
+			input:        makeSimpleTwapInput(baseTimePlusKeepPeriod.Add(-time.Millisecond), baseTimePlusKeepPeriod.Add(-time.Millisecond), quoteAssetA),
+			expTwap:      sdk.NewDec(10),
+		},
+		"(1 record younger than keep threshold); to now; ctxTime = start time = end time = after keep threshold": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      oneHourAfterKeepThreshold,
+			input:        makeSimpleTwapInput(oneHourAfterKeepThreshold, oneHourAfterKeepThreshold, quoteAssetA),
+			expTwap:      sdk.NewDec(10),
+		},
+		"(1 record younger than keep threshold); with end time; ctxTime = start time = end time = after keep threshold - 1ns": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      oneHourAfterKeepThreshold,
+			input:        makeSimpleTwapInput(oneHourAfterKeepThreshold.Add(-time.Millisecond), oneHourAfterKeepThreshold.Add(-time.Millisecond), quoteAssetA),
```

Comment:
I feel like we could probably add some of these table cases more clearly, via a for loop. Don't need to change if you don't want to, but I think could make adding more cases easier

Reply:
I explored this idea [here](https://github.com/osmosis-labs/osmosis/pull/2543/commits/e8e5084271c680ae2310f830ce4b8df1fc565968)

I don't think this is easily doable because if I generalize test cases to always subtract from end time to trigger the "with end time" path, the expected value of `expTwap` changes.

I cannot generalize by instead modifying `ctxTime` to trigger the "with end time" path because test cases' logic sometimes depends on it.

Please let me know if you see something I'm not, and I can change this